src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutBindings.cs

```diff
@@ -0,0 +1,378 @@
+		{
+			List<ITaskItem> resourceItems = group?.Items;
+			if (resourceItems == null || resourceItems.Count == 0)
+				return;
+
+			ITaskItem item = resourceItems.FirstOrDefault ();
+			if (item == null)
+				return;
+
+			string collectionKey;
+			if (!GetRequiredMetadata (item, CalculateLayoutCodeBehind.WidgetCollectionKeyMetadata, out collectionKey))
+				return;
+
+			ICollection<LayoutWidget> widgets;
+			if (!CalculateLayoutCodeBehind.LayoutWidgets.TryGetValue (collectionKey, out widgets) || widgets.Count == 0) {
+				string inputPaths = String.Join ("; ", resourceItems.Select (i => i.ItemSpec));
+				LogWarning ($"No widgets found for layout ({inputPaths})");
+				return;
+			}
+
+			string outputFileName;
+			if (!GetRequiredMetadata (item, CalculateLayoutCodeBehind.LayoutBindingFileNameMetadata, out outputFileName))
+				return;
+
+			string fullClassName;
+			if (!GetRequiredMetadata (item, CalculateLayoutCodeBehind.ClassNameMetadata, out fullClassName))
+				return;
+
+			string outputFilePath = Path.Combine (MonoAndroidCodeBehindDir, outputFileName);
+			string classNamespace;
+			string className;
+			int idx = fullClassName.LastIndexOf ('.');
+			if (idx < 0) {
+				classNamespace = String.Empty;
+				className = fullClassName;
+			} else {
+				bool fail = false;
+				classNamespace = fullClassName.Substring (0, idx);
+				if (String.IsNullOrEmpty (classNamespace)) {
+					LogError ($"Malformed full class name '{fullClassName}'. Missing namespace.");
+					fail = true;
+				}
+
+				className = fullClassName.Substring (idx + 1);
+				if (String.IsNullOrEmpty (className)) {
+					LogError ($"Malformed full class name '{fullClassName}'. Missing class name.");
+					fail = true;
+				}
+
+				if (fail)
+					return;
+			}
+
+			if (!GenerateSource (generator, outputFilePath, widgets, classNamespace, className, group.Classes))
+				return;
+
+			pathAdder (outputFilePath);
+		}
+
+		bool GenerateSource (BindingGenerator generator, string outputFilePath, ICollection <LayoutWidget> widgets, string classNamespace, string className, List<PartialClass> partialClasses)
+		{
+			try {
+				if (partialClasses != null && partialClasses.Count > 0) {
+					foreach (var pc in partialClasses) {
+						if (pc == null)
+							continue;
+
+						pc.Stream = File.Open (pc.OutputFilePath, FileMode.Create);
+						pc.Writer = new StreamWriter (pc.Stream, Encoding.UTF8);
+					}
+				}
+
+				using (var fs = File.Open (outputFilePath, FileMode.Create)) {
+					using (var sw = new StreamWriter (fs, Encoding.UTF8)) {
+						return GenerateSource (sw, generator, widgets, classNamespace, className, partialClasses);
+					}
+				}
+			} finally {
+				if (partialClasses != null && partialClasses.Count > 0) {
+					foreach (var pc in partialClasses) {
+						if (pc == null)
+							continue;
+
+						if (pc.Writer != null) {
+							pc.Writer.Close ();
+							pc.Writer.Dispose ();
```

Comment:
You shouldn't need both `.Close()` and `.Dispose()`.

Reply:
In theory, yes. The operative word is *shouldn't*. Sometimes you must sometimes you don't have to, depending on the implementation of the writer, the underlying stream, how the writer was created (does it own the stream or not?) and I don't want to guess on those things - having both calls doesn't hurt and works in all cases, no matter how the other code changes.